distributed/comm/mpi.py

```diff
@@ -0,0 +1,781 @@
+                tag = tag + 1
+
+                # Send which frames are CUDA (bool) and
+                # how large each frame is (uint64)
+                #await self.ep.send(
+                #    struct.pack(nframes * "?" + nframes * "Q", *cuda_frames, *sizes)
+                #)
+                packed_msg = struct.pack(nframes * "?" + nframes * "Q", \
+                                                         *cuda_frames, *sizes)
+                s_size = struct.calcsize(nframes * "?" + nframes * "Q")
+
+                await self.mpi_send(packed_msg, s_size, tag)
+                tag = tag + 1
+                logger.debug("write_2: me=%s, you=%s, tag=%s", MPI.COMM_WORLD.Get_rank(), self.peer_rank, tag)
+
+                # Send frames
+
+                # It is necessary to first synchronize the default stream 
+                # before start sending. 
+
+                # non-blocking CUDA streams.
+                if any(cuda_send_frames):
+                    synchronize_stream(0)
+
+                sizes_new = tuple(i for i in sizes if i != 0) # remove all 0's from this list
+                read_counter = 0
+
+                for each_frame in send_frames:
+                    #await self.ep.send(each_frame)
+                    s_size = sizes_new[read_counter]
+                    await self.mpi_send(each_frame, s_size, tag)
+                    tag = tag + 1
+                    read_counter = read_counter + 1
+                logger.debug("write_3: me=%s, you=%s, tag=%s, sizes=%s, sizes_new=%s", MPI.COMM_WORLD.Get_rank(), self.peer_rank, tag, sizes, sizes_new)
+                return sum(sizes_new)
+            except (Exception):
+                self.abort()
+                raise CommClosedError("While writing, the connection was closed")
+
+    # a utility function for sending messages larger than CHUNK_SIZE data
+    async def mpi_send_large(self, buf, size, _tag):
+
+        me = MPI.COMM_WORLD.Get_rank()
+        you = self.peer_rank
+         
+        logger.debug("mpi_send_large: host=%s, me=%s, you=%s, tag=%s, size=%s, type(buf)=%s", \
+                                               socket.gethostname(), me, you, _tag, size, type(buf))
+
+        blk_size = CHUNK_SIZE
+        num_of_blks   = int(size / blk_size)
+        last_blk_size =     size % blk_size
+
+        logger.debug("mpi_send_large: blk_size=%s, num_of_blks=%s, last_blk_size=%s", \
+                                                          blk_size, num_of_blks, last_blk_size)
+
+        num_of_reqs = num_of_blks
+
+        if last_blk_size is not 0:
+            num_of_reqs = num_of_reqs + 1
+
+        reqs = []        
+
+        for i in range(num_of_blks):
+            s_idx = (i)   * blk_size
+            e_idx = (i+1) * blk_size
+
+            if type(buf) == rmm._lib.device_buffer.DeviceBuffer:
+                # need this if because rmm.DeviceBuffer is not subscriptable
+                shadow_buf = rmm.DeviceBuffer(ptr=(buf.ptr+s_idx), \
```

Comment:
Any reason not to use `device_array` here as that is defined above?

Reply:
Same reason as above. If there's a suggestion to improve, please let us know.

@ZackWRyan

Reply:
To add to what Aamir said, to replace subscripting/slicing for RMM buffers it's necessary to create these 'shadow' buffers that are actually just chunks of the larger RMM buffer, and in order to do that without copying we need to specify a pointer to an intermediate spot in the buffer.  `device_array` doesn't allow us to specify a pointer, only a size, partially because that `device_array` might actually be a Numba array (not RMM), and as far as I can tell, Numba doesn't support creating device buffers at specific pointer locations.